### PR TITLE
Metrics for streamed elements

### PR DESCRIPTION
**Pull request by mziolekda**
Add metrics counting the elements in transaction, transaction-tree, completion and acs streams

### CODE REVIEW

**Review by gerolf-da — APPROVED** [review]
Thanks, @mziolekda!

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiActiveContractsService.scala [hunk 1]
```diff
@@ -42,6 +46,10 @@ private[apiserver] final class ApiActiveContractsService private (
           .validate(request.getFilter)
           .fold(Source.failed, backend.getActiveContracts(_, request.verbose))
           .via(logger.logErrorsOnStream)
+          .map(item => {
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiActiveContractsService.scala:49]: Counting for all consumers might lead to metrics with ambiguous metrics when we have multiple consumers with varying behaviors in the application. Are you ok with this limitation or should we rather particularize the metric per consumer somehow (e.g. use the traceId)?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/transaction/ApiTransactionService.scala [hunk 2]
```diff
@@ -115,6 +125,10 @@ private[apiserver] final class ApiTransactionService private (
         )
         .via(logger.debugStream(transactionTreesLoggable))
         .via(logger.logErrorsOnStream)
+        .map(item => {
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/transaction/ApiTransactionService.scala:128]: I would extract a generic/extension method and instrument all sources with it for conciseness

Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/transaction/ApiTransactionService.scala:128]: e.g. 
```
  implicit class InstrumentedCountingSource[E, N](s: Source[E, N]) {
    def counter(counter: Counter): Source[E, N] =
      s.map(item => {
        counter.inc()
        item
      })
  }
```
or a simple object method if extension methods don't appeal to you. What do you think?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiActiveContractsService.scala (hunk 1 above)
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiActiveContractsService.scala:49]: You cannot expand the metrics namespace without serious consequences for the metrics gathering systems. Wavefront does not support well a dynamically growing metrics namespace (datadog also does not), we are also reaching the limit of what grafana is willing to accept. 
The solution would be to add labels, in dropwizard terms they are called tags. They will be supported in version 5.0 of dropwizard, which has been i n the making for the past 3 years.

**Review by tudor-da — APPROVED** [review]
LGTM!

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiActiveContractsService.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiActiveContractsService.scala]: You're right that it's not ideal to have dynamic metrics namespace, but we already do have them in a lot of places. If this solution serves the need now, let's leave it like this and keep metric specialization in mind for future iterations.

**Pull request merged**